From 4da861368b88c03b22993f95de5b508ad5637c25 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pavel=20B=C5=99ezina?= <pbrezina@redhat.com>
Date: Tue, 11 Oct 2022 12:30:56 +0200
Subject: [PATCH 2/6] monitor: read all enabled domains in
 add_implicit_services
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Reading sssd/domains option is no longer sufficient since domains
can be enabled through domain/enabled.

Reviewed-by: Alejandro López <allopez@redhat.com>
Reviewed-by: Sumit Bose <sbose@redhat.com>
---
 src/confdb/confdb.c   | 15 ++-------------
 src/confdb/confdb.h   | 11 +++++++++++
 src/monitor/monitor.c |  9 +++++----
 3 files changed, 18 insertions(+), 17 deletions(-)

diff --git a/src/confdb/confdb.c b/src/confdb/confdb.c
index 9465bffe394ebed783b8217f96049f3d07ba7e77..3ecdaa3b91e3b550f670768d336e9d7d85bb66e6 100644
--- a/src/confdb/confdb.c
+++ b/src/confdb/confdb.c
@@ -45,9 +45,6 @@
 /* SSSD domain name that is used for the auto-configured files domain */
 #define IMPLICIT_FILES_DOMAIN_NAME "implicit_files"
 
-
-static int confdb_get_enabled_domain_list(struct confdb_ctx *cdb,
-                                          TALLOC_CTX *ctx, char ***_result);
 static int confdb_get_domain_enabled(struct confdb_ctx *cdb,
                                      const char *domain, bool *_enabled);
 
@@ -2675,16 +2672,8 @@ done:
     return ret;
 }
 
-/**
- * Retrieve the list of enabled domains considering the explicit list
- * and the 'enabled' attribute.
- * @param cdb The database configuration context.
- * @param ctx The memory context.
- * @param result Output variable where the list of domains will be stored.
- * @return 0 if the list was retrieved properly, another value on error.
- */
-static int confdb_get_enabled_domain_list(struct confdb_ctx *cdb,
-                                          TALLOC_CTX *ctx, char ***_result)
+int confdb_get_enabled_domain_list(struct confdb_ctx *cdb,
+                                   TALLOC_CTX *ctx, char ***_result)
 {
     int ret;
     char **domlist = NULL;
diff --git a/src/confdb/confdb.h b/src/confdb/confdb.h
index 0effd1193708676ade230d8922ab78fae9b5c15a..a53894846737467f12e5dcd99ce35ebee165ccf7 100644
--- a/src/confdb/confdb.h
+++ b/src/confdb/confdb.h
@@ -497,6 +497,17 @@ int confdb_get_domain(struct confdb_ctx *cdb,
 int confdb_get_domains(struct confdb_ctx *cdb,
                        struct sss_domain_info **domains);
 
+/**
+ * Retrieve the list of enabled domains considering the explicit list
+ * and the 'enabled' attribute.
+ * @param cdb The database configuration context.
+ * @param ctx The memory context.
+ * @param result Output variable where the list of domains will be stored.
+ * @return 0 if the list was retrieved properly, ENOENT if no domain is enabled, another value on error.
+ */
+int confdb_get_enabled_domain_list(struct confdb_ctx *cdb,
+                                   TALLOC_CTX *ctx, char ***_result);
+
 int confdb_expand_app_domains(struct confdb_ctx *cdb);
 
 /**
diff --git a/src/monitor/monitor.c b/src/monitor/monitor.c
index 7670114d37646ebcacd1d0f8c6876e40ff03938e..511e13971d253bc25cb9a04008c1a363f2182748 100644
--- a/src/monitor/monitor.c
+++ b/src/monitor/monitor.c
@@ -777,13 +777,14 @@ static errno_t add_implicit_services(struct confdb_ctx *cdb, TALLOC_CTX *mem_ctx
         return ENOMEM;
     }
 
-    ret = confdb_get_string_as_list(cdb, tmp_ctx,
-                                    CONFDB_MONITOR_CONF_ENTRY,
-                                    CONFDB_MONITOR_ACTIVE_DOMAINS,
-                                    &domain_names);
+    ret = confdb_get_enabled_domain_list(cdb, tmp_ctx, &domain_names);
     if (ret == ENOENT) {
         DEBUG(SSSDBG_OP_FAILURE, "No domains configured!\n");
         goto done;
+    } else if (ret != EOK) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Error retrieving domains list [%d]: %s\n",
+              ret, sss_strerror(ret));
+        goto done;
     }
 
     ret = confdb_get_bool(cdb, CONFDB_MONITOR_CONF_ENTRY,
-- 
2.37.3

